Tutorial chapter about configuring FastDDS middleware with XML files (#1102)
ros2/ros2_documentation: source/Tutorials/FastDDS-Configuration/FastDDS-Configuration.rst

Open the ``SyncAsync.xml`` file and add the following profiles inside the ``<profiles>`` tag:

.. code-block:: XML

    <!-- subscriber profile for topic sync_topic -->
    <subscriber profile_name="default_subscriber" is_default_profile="true">
        <historyMemoryPolicy>DYNAMIC</historyMemoryPolicy>
    </subscriber>

Execute the subscriber node
^^^^^^^^^^^^^^^^^^^^^^^^^^^

With the publisher node running in one terminal, open another one and export the required environment variables for the XML to be loaded:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

      export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
      export RMW_FASTRTPS_USE_QOS_FROM_XML=1
      export FASTRTPS_DEFAULT_PROFILES_FILE=path/to/SyncAsync.xml

  .. group-tab:: macOS

    .. code-block:: console

      export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
      export RMW_FASTRTPS_USE_QOS_FROM_XML=1
      export FASTRTPS_DEFAULT_PROFILES_FILE=path/to/SyncAsync.xml

  .. group-tab:: Windows

    .. code-block:: console

      SET RMW_IMPLEMENTATION=rmw_fastrtps_cpp
      SET RMW_FASTRTPS_USE_QOS_FROM_XML=1
      SET FASTRTPS_DEFAULT_PROFILES_FILE=path/to/SyncAsync.xml

> Finally, ensure you have sourced your setup files and run the node:

Author: I added the sourcing command to all the command-blocks in the pull request

        </matchedSubscribersAllocation>
    </publisher>

The number of matching subscribers is being limited to one.

Now open three terminals and do not forget to source the setup files and to set the required environment variables.
On the first terminal run the publisher node, and the subscriber node on the other two.
You should see that only the first subscriber node is receiving the messages from both topics.
The second one could not complete the matching process in the ``/async_topic`` because the publisher prevented it, as it had already reached its maximum of matched publishers.
Consequently, only the messages from the ``/sync_topic`` are going to be received in this third terminal:

.. code-block:: console

    [INFO] [1613127657.088860890] [sync_async_subscriber]: I heard: 'SYNC: Hello, world! 18'
    [INFO] [1613127657.588896594] [sync_async_subscriber]: I heard: 'SYNC: Hello, world! 19'
    [INFO] [1613127658.088849401] [sync_async_subscriber]: I heard: 'SYNC: Hello, world! 20'

Using partitions within the topic
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The partitions feature can be used to control which publishers and subscribers exchange information within the same topic.

> Let us change the ``/sync_topic`` publisher to partition ``part1`` and create a new ``/sync_topic`` subscriber which uses partition ``part2``.

Reviewer: I think an example of when and why you want to use this feature would be helpful.

.. code-block:: console

    [INFO] [1612972054.995410057] [sync_async_subscriber]: I heard: 'ASYNC: Hello, world! 10'
    [INFO] [1612972055.995396561] [sync_async_subscriber]: I heard: 'ASYNC: Hello, world! 11'
    [INFO] [1612972056.995473953] [sync_async_subscriber]: I heard: 'ASYNC: Hello, world! 12'

Configuring a service and a client
----------------------------------

Services and clients have a publisher and a subscriber each, that communicate through two different topics.
For example, for a service named ``ping`` there is:

* A service subscriber listening to requests on ``/rq/ping``.
* A service publisher sending responses on ``/rr/ping``.
* A client publisher sending requests on ``/rq/ping``.
* A client subscriber listening to responses on ``/rr/ping``.

Although you can use these topic names to set the configuration profiles on the XML, sometimes you may wish to apply the same profile to all services or clients on a node.
Instead of copying the same profile with all topic names generated for all services, you can just create a publisher and subscriber profile pair named ``service``.

> The same can be done for clientes creating a pair named ``client``.

Reviewer: Clientes => Clients

The interface between the ROS 2 stack and *Fast DDS* is provided by the ROS 2 middleware implementation `rmw_fastrtps <https://github.com/ros2/rmw_fastrtps>`_.
This implementation is available in all ROS 2 distributions, both from binaries and from sources.

ROS 2 RMW only allows for the configuration of certain middleware QoS
(see `ROS 2 QoS policies <https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/#qos-policies>`_).
However, ``rmw_fastrtps`` offers extended configuration capabilities
to take full advantage of the features in *Fast DDS*.
This tutorial will guide you through a series of examples explaining how to use XML files to unlock this extended configuration.

In order to get more information about using *Fast DDS* on ROS 2, please check the `following documentation <https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/ros2.html>`__.

Prerequisites
-------------

This tutorial does not work with ROS 2 Foxy.
For this tutorial to work with ROS 2 Foxy the [ros2/rmw_fastrtps#335](https://github.com/ros2/rmw_fastrtps/pull/335) and [ros2/rmw_fastrtps#497](https://github.com/ros2/rmw_fastrtps/pull/497) should be backported to the foxy branch.

> Right now, this tutorial only works in ROS 2 Rolling and newer installations.

Reviewer 2: We're now switching to one-branch-per-distribution.  With that in mind, I've retargeted this PR to the 'rolling' branch.  Thus, this note isn't actually needed anymore; we will only show this tutorial on rolling.

   :depth: 2
   :local:

Background
----------

The interface between the ROS 2 stack and *Fast DDS* is provided by the ROS 2 middleware implementation `rmw_fastrtps <https://github.com/ros2/rmw_fastrtps>`_.
This implementation is available in all ROS 2 distributions, both from binaries and from sources.

ROS 2 RMW only allows for the configuration of certain middleware QoS
(see `ROS 2 QoS policies <https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/#qos-policies>`_).
However, ``rmw_fastrtps`` offers extended configuration capabilities
to take full advantage of the features in *Fast DDS*.
This tutorial will guide you through a series of examples explaining how to use XML files to unlock this extended configuration.

In order to get more information about using *Fast DDS* on ROS 2, please check the `following documentation <https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/ros2.html>`__.

Prerequisites
-------------

This tutorial assumes that you know how to :ref:`create a package <CreatePkg>`.
It also assumes you know how to write a :ref:`simple publisher and subscriber<CppPubSub>` and a :ref:`simple service and client<CppSrvCli>`.

> Although the examples are implemented on C++, the same concepts apply to Python packages.

Reviewer 2: ```suggestion
Although the examples are implemented in C++, the same concepts apply to Python packages.
```

Mixing synchronous and asynchronous publications in the same node
-----------------------------------------------------------------

> On this first example, a node with two publishers, one of them with synchronous publication mode and the other one with asynchronous publication mode, will be created.

Reviewer 2: ```suggestion
In this first example, a node with two publishers, one of them with synchronous publication mode and the other one with asynchronous publication mode, will be created.
```

``rmw_fastrtps`` uses asynchronous publication mode by default. When the publisher invokes the write operation, the data is copied into a queue,
a background thread (asynchronous thread) is notified about the addition to the queue, and control of the thread is returned to the user before the data is actually sent.
The background thread is in charge of consuming the queue and sending the data to every matched reader.

On the other hand, with synchronous publication mode the data is sent directly within the context of the user thread.
This entails that any blocking call occurring during the write operation would block the user thread, thus preventing the application from continuing its operation.
However, his mode typically yields higher throughput rates at lower latencies, since there is no notification nor context switching between threads.

Create the node with the publishers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Start creating the node with both publishers.

> First, create a new package named ``sync_async_node_example_cpp``.

Reviewer 2: These two lines seem redundant.  Also, it would be nice to give the commands to do this, something like:

```
mkdir -p fastdds_ws/src
cd fastdds_ws/src
ros2 pkg create --build-type ament_cmake --dependencies rclcpp std_msgs sync_async_node_example_cpp
```

> Then, add a file named ``sync_async_writer.cpp`` to thepackage, with the following content.

Reviewer 2: ```suggestion
Then, add a file named ``src/sync_async_writer.cpp`` to the package, with the following content.
```

            // Publish the message using the synchronous publisher
            sync_publisher_->publish(sync_message);

            // Log the message to the console to show progress
            RCLCPP_INFO(this->get_logger(), "Asynchronously publishing: '%s'", async_message.data.c_str());

            // Publish the message using the asynchronous publisher
            async_publisher_->publish(async_message);

            // Prepare the count for the next message
            count_++;
        }

        // This timer will trigger the publication of new data every half a second
        rclcpp::TimerBase::SharedPtr timer_;

        // A publisher that publishes asynchronously
        rclcpp::Publisher<std_msgs::msg::String>::SharedPtr async_publisher_;

        // A publisher that publishes synchronously
        rclcpp::Publisher<std_msgs::msg::String>::SharedPtr sync_publisher_;

        // Number of messages sent so far
        size_t count_;
    };

    int main(int argc, char * argv[])
    {
        rclcpp::init(argc, argv);
        rclcpp::spin(std::make_shared<SyncAsyncPublisher>());
        rclcpp::shutdown();
        return 0;
    }

> Add a new target on the ``CMakeLists.txt`` file of your package with this code and name it ``SyncAsyncWriter``, so that you can run the node using ``ros2 run``.

Reviewer 2: It would be much easier for the reader if we gave an example of what the `CMakeLists.txt` code should look like.

If this node is built and run now, both publishers will behave the same, publishing asynchronously in both topics, because this is the default publication mode.

> The default publication mode configuration can be changed on runtime during the node launching, using an XML file.

Reviewer 2: ```suggestion
The default publication mode configuration can be changed in runtime during the node launching, using an XML file.
```

.. code-block:: console

    source install/setup.bash
    ros2 run sync_async_node_example_cpp SyncAsyncWriter

You should see the publishers sending the data from the publishing node, like so:

.. code-block:: console

    [INFO] [1612972049.994630332] [sync_async_publisher]: Synchronously publishing: 'SYNC: Hello, world! 0'
    [INFO] [1612972049.995097767] [sync_async_publisher]: Asynchronously publishing: 'ASYNC: Hello, world! 0'
    [INFO] [1612972050.494478706] [sync_async_publisher]: Synchronously publishing: 'SYNC: Hello, world! 1'
    [INFO] [1612972050.494664334] [sync_async_publisher]: Asynchronously publishing: 'ASYNC: Hello, world! 1'
    [INFO] [1612972050.994368474] [sync_async_publisher]: Synchronously publishing: 'SYNC: Hello, world! 2'
    [INFO] [1612972050.994549851] [sync_async_publisher]: Asynchronously publishing: 'ASYNC: Hello, world! 2'

Now you have a synchronous publisher and an asynchronous publisher running inside the same node.

Create a node with the subscribers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Next, a new node with the subscribers that will listen to the ``sync_topic`` and ``async_topic`` publications is going to be created.

> In a new source file named ``sync_async_reader.cpp`` write the following content:

Reviewer 2: ```suggestion
In a new source file named ``src/sync_async_reader.cpp`` write the following content:
```

                "sync_topic", 10, std::bind(&SyncAsyncSubscriber::topic_callback, this, _1));

            // Create the asynchronous subscriber on topic 'async_topic'
            // and tie it to the topic_callback
            async_subscription_ = this->create_subscription<std_msgs::msg::String>(
                "async_topic", 10, std::bind(&SyncAsyncSubscriber::topic_callback, this, _1));
        }

    private:

        /**
         * Actions to run every time a new message is received
         */
        void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
        {
            RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg->data.c_str());
        }

        // A subscriber that listens to topic 'sync_topic'
        rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sync_subscription_;

        // A subscriber that listens to topic 'async_topic'
        rclcpp::Subscription<std_msgs::msg::String>::SharedPtr async_subscription_;
    };

    int main(int argc, char * argv[])
    {
        rclcpp::init(argc, argv);
        rclcpp::spin(std::make_shared<SyncAsyncSubscriber>());
        rclcpp::shutdown();
        return 0;
    }

> Add a new target on the ``CMakeLists.txt`` file linked to this code and name it ``SyncAsyncReader``.

Reviewer 2: Same as earlier; it would be much nicer if there was example `CMakeLists.txt` code here.

and ``rmw_fastrtps`` will apply this profile to all publishers and subscribers for that topic.
The default configuration profile is identified by the attribute ``is_default_profile`` set to ``true``, and acts as a fallback profile when there is no other one with a name matching the topic name.

The environment variable ``FASTRTPS_DEFAULT_PROFILES_FILE`` is used to inform *Fast DDS* the path to the XML file with the configuration profiles to load.

RMW_FASTRTPS_USE_QOS_FROM_XML
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Among all the configurable attributes, ``rmw_fastrtps`` treats ``publishMode`` and ``historyMemoryPolicy`` differently.
By default, these values are set to ``ASYNCHRONOUS`` and ``PREALLOCATED_WITH_REALLOC`` within the ``rmw_fastrtps`` implementation, and the values set on the XML file are ignored.
In order to use the values in the XML file, the environment variable ``RMW_FASTRTPS_USE_QOS_FROM_XML`` must be set to ``1``.

However, this entails **another caveat**: If ``RMW_FASTRTPS_USE_QOS_FROM_XML`` is set, but the XML file does not define
``publishMode`` or ``historyMemoryPolicy``, these attributes take the *Fast DDS* default value instead of the ``rmw_fastrtps`` default value.

> This is important, specially for ``historyMemoryPolicy``, because the *Fast DDS* deafult value is ``PREALLOCATED`` which does not work with ROS2 topic data types.

Reviewer 2: ```suggestion
This is important, especially for ``historyMemoryPolicy``, because the *Fast DDS* deafult value is ``PREALLOCATED`` which does not work with ROS2 topic data types.
```

Limiting the number of matching subscribers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Add a maximum number of matched subscribers to the ``/async_topic`` publisher profile.
It should look like this:

.. code-block:: XML

    <!-- publisher profile for topic async_topic -->
    <publisher profile_name="/async_topic">
        <historyMemoryPolicy>DYNAMIC</historyMemoryPolicy>
        <qos>
            <publishMode>
                <kind>ASYNCHRONOUS</kind>
            </publishMode>
        </qos>
        <matchedSubscribersAllocation>
            <initial>0</initial>
            <maximum>1</maximum>
            <increment>1</increment>
        </matchedSubscribersAllocation>
    </publisher>

The number of matching subscribers is being limited to one.

Now open three terminals and do not forget to source the setup files and to set the required environment variables.
On the first terminal run the publisher node, and the subscriber node on the other two.

> You should see that only the first subscriber node is receiving the messages from both topics.

Reviewer 2: ```suggestion
You should see that only the first subscriber node receives the messages from both topics.
```

    [INFO] [1612972055.995396561] [sync_async_subscriber]: I heard: 'ASYNC: Hello, world! 11'
    [INFO] [1612972056.995473953] [sync_async_subscriber]: I heard: 'ASYNC: Hello, world! 12'

Configuring a service and a client
----------------------------------

Services and clients have a publisher and a subscriber each, that communicate through two different topics.
For example, for a service named ``ping`` there is:

* A service subscriber listening to requests on ``/rq/ping``.
* A service publisher sending responses on ``/rr/ping``.
* A client publisher sending requests on ``/rq/ping``.
* A client subscriber listening to responses on ``/rr/ping``.

Although you can use these topic names to set the configuration profiles on the XML, sometimes you may wish to apply the same profile to all services or clients on a node.
Instead of copying the same profile with all topic names generated for all services, you can just create a publisher and subscriber profile pair named ``service``.
The same can be done for clients creating a pair named ``client``.

Create the nodes with the service and client
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Start creating the node with the service.

> Add a new source file named ``ping_service.cpp`` on your package with the following content:

Reviewer 2: ```suggestion
Add a new source file named ``src/ping_service.cpp`` on your package with the following content:
```
